nx/lib/nx/random.ex

```diff
@@ -0,0 +1,475 @@
+defmodule Nx.Random do
+  @moduledoc """
+
+  ## PRNG (Pseudo-Random Number Generator)
+
+  Unlike the stateful pseudo-random number
+  generators (PRNGs) that users of most programming
+  languages and numerical libraries
+  may be accustomed to, Nx random functions all
+  require an explicit PRNG state to be passed as
+  a first argument. The random state is described by
+  an `Nx.Tensor` composed by 2 unsigned 32-bit integers
+  that we call a key, usually generated by the `Nx.Random.key/1`
+  function:
+
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+
+  This key can then be used in any of
+  Nx’s random number generation routines:
+
+      iex> key = Nx.Random.key(12)
+      iex> Nx.Random.uniform(key)
+      #Nx.Tensor<
+        f32
+        0.3037703037261963
+      >
+
+  Note that using a key does not modify it, so
+  reusing the same key will lead to the same result:
+
+      iex> key = Nx.Random.key(12)
+      iex> Nx.Random.uniform(key)
+      #Nx.Tensor<
+        f32
+        0.3037703037261963
+      >
+
+  If you need a new random number, you
+  can use `Nx.Random.split/2` to generate new subkeys:
+
+      iex> key = Nx.Random.key(12)
+      iex> keys = Nx.Random.split(key)
+      iex> Nx.Random.uniform(keys[0])
+      #Nx.Tensor<
+        f32
+        0.1531379222869873
+      >
+      iex> Nx.Random.uniform(keys[1])
+      #Nx.Tensor<
+        f32
+        0.7691127061843872
+      >
+
+  ## Advanced
+
+  Design and Context
+
+   In short, Nx's PRNGs are based on a Threefry counter PRNG
+   associated to a functional array-oriented splitting model
+
+  To summarize, among other requirements, Nx's PRNG aims to:
+
+  1. Ensure reproducibility;
+
+  2. Parallelize well, both in terms of vectorization (generating array values)
+     and multi-replica, multi-core computation. In particular it should not use
+     sequencing constraints between random function calls.
+  """
+
+  import Nx.Defn, only: [deftransformp: 2, defn: 2, defnp: 2]
+
+  @nbits 32
+
+  @doc """
+  Create a pseudo-random number generator (PRNG) key given an integer seed.
+
+  ## Examples
+
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+
+      iex> Nx.Random.key(999999999999)
+      #Nx.Tensor<
+        u32[2]
+        [232, 3567587327]
+      >
+  """
+  defn key(seed) do
+    k1 = Nx.right_shift(seed, 32)
+    k2 = Nx.bitwise_and(seed, 0xFFFFFFFF)
+
+    Nx.stack([k1, k2])
+    |> Nx.as_type({:u, 32})
+  end
```

Comment:
I don't quite get the question? Clause in random_bits has nothing to do with key length, it corresponds with the requested bits number.

Reply:
I meant that while the rest of the API accepts type parameters, this implementation only seems to work by receiving a 64bit value and outputting u32. Perhaps we should explicitly upcast u8, u16 and u32 to u64 and s8, s16 and s32 to s64 before passing into `right_shift` and `bitwise_and`

Otherwise, from your comment I understand that the key width doesn't change regardless of the output type and width for the generator. Is this right?

Reply:
That is right, key width always stay the same.

Is the issue that implementation SEEMS to work only for 64bit seed or that it WORKS only for 64bit (which isn't true)? We can, of course, explicitly cast them if that improves readability.

Reply:
I think we should raise if the input seed has less than 64 bits to ensure we have the expected entropy in the keyset generated

Reply:
What keyset? We get entropy from the threefry function, key is literally 64bit seed divided in half or zero and seed if it's 32bit or smaller. Now I'm even more confused?

Reply:
What I mean is that the `key` function expects to output 2 useful numbers k1 and k2, ultimately.
However, if the input has 32 bits k1 will always be 0 and k2 will always be the seed. There's less information being output than what would otherwise be expected. For 16 and 8 bits this is a bigger problem.

Just to make sure we're on the same page, we have to keep in mind that `seed` is a scalar Nx.Tensor, not a plain number, even if you pass a plain number as the input.

Given all of this, I'm not sure if having less useful information in the seed will affect the algorithm's output that much. This is why I think it's probably best that we enforce the `seed` to be given as a 64bit input

Reply:
The thing is, it's exactly what is expected. Even 63 consecutive 0s are as valuable for us as any other number. There's never "less" information in that context. And it won't affect the algorithm at all. Even if we enforce 64bit input there is no rule that invalidates input like 1 or 2 which are essentialy 2bit numbers. We simply don't care if we get 63 0sand 1 or 7 0s and 1 because we either way get the same amount of information. In both cases we'll get [0,1]. I think we are misunderstanding something

Reply:
I get what you mean. If you say the algorithm won't be affected by having less information (which there is because we have half the useful bits and therefore 2ˆ32 less keys to choose from), we can move forward ignoring this

Reply:
We can safely ignore this as this is something entirely up to user's decision. Doesn't matter what seed he provides, we can split up to 2^64 keys regardless.